codersdk/wsjson/decoder.go

```diff
@@ -0,0 +1,74 @@
+package wsjson
+
+import (
+	"context"
+	"encoding/json"
+	"sync/atomic"
+
+	"nhooyr.io/websocket"
+
+	"cdr.dev/slog"
+)
+
+type Decoder[T any] struct {
+	conn       *websocket.Conn
+	typ        websocket.MessageType
+	ctx        context.Context
+	cancel     context.CancelFunc
+	chanCalled atomic.Bool
+	logger     slog.Logger
+}
+
+// Chan starts the decoder reading from the websocket and returns a channel for reading the
+// resulting values. The chan T is closed if the underlying websocket is closed, or we encounter an
+// error. We also close the underlying websocket if we encounter an error reading or decoding.
+func (d *Decoder[T]) Chan() <-chan T {
+	if d.chanCalled.Load() {
+		panic("chan called more than once")
+	}
+	d.chanCalled.Store(true)
+	values := make(chan T)
```

Comment:
```suggestion
	values := make(chan T, 1)
```
Suggestion: Performance, allow us to start reading the next message instead of waiting for the current one to be grabbed first.